### PR TITLE
[source-hubspot] - Transform Deals and Contacts stream "new" fields to legacy fields

### PR DESCRIPTION
**Pull request by pnilan**
## What
- On Nov 20, 2024, HubSpot will be sunsetting the following fields from the Contacts and Deals (and related) streams per [HubSpot Changelog](https://developers.hubspot.com/changelog/upcoming-sunset-of-pipeline-legacy-stage-calculated-properties-for-contacts-and-deals):
    - hs_lifecyclestage_{stage id}
    - hs_date_entered_{stage id}
    - hs_date_exited_{stage id}
    - hs_time_in_{stage id}

And replace them with the following streams:
- hs_v2_date_entered_{stage_id}
- hs_v2_date_exited_{stage_id}
- hs_v2_latest_time_in_{stage_id}
- hs_v2_cumulative_time_in_{stage_id}

Note that that `hs_v2_cumulative_time_in_{stage_id}` is actually a new field and therefore does not need to be transformed for backwards compatibility. Additionally, per this [hubspot community thread](https://community.hubspot.com/t5/Developer-Announcements/Upcoming-Sunset-of-Legacy-Stage-Calculated-Properties-for/m-p/1026312#M790), the legacy-equivalent of `hs_v2_date_entered_{stage_id}` is both `hs_lifecyclestage_{stage_id}_date` *and* `hs_date_enteredd_{stage_id}`.

This PR implements the logic necessary to avoid a breaking change for users by automatically adding the legacy fields into the dynamically generated schema and into each record. This enables users' syncs to continue without any action (e.g. schema refresh or stream refresh).

## How
- New `LegacyFieldTransformation` custom component which inherits from `RecordTransformation`
    - Will be compatible with low-code as a custom component when the source is migrated
- New `LegacyFieldMigrationMixin`class which overrides the `properties` property and the `_transform` method to invoke the parent class' methods and then execute the new transformation logic.
    - While overriding a private method was not the preferred solution, the `_transform` method had already been overridden multiple times within various streams and would enable easy implementation of this new transformation logic.
    - Additionally, when the source is migrated to low-code, it will be easy to remove this logic and directly reference the new `LegacyFieldTransformation` component within the YAML file.

## Review guide
1. `components.py`
2. `streams.py`
3. `test_components.py`
4. `test_streams.py`

## User Impact
None -- that was the goal!

## Can this PR be safely reverted and rolled back?
- [ ] YES 💚
- [X] NO ❌

### CODE REVIEW

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/components.py
Comment (maxi297) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/components.py:26]: The return type differs from the parent https://github.com/airbytehq/airbyte-python-cdk/blob/39786d26926d8b041df5d7b6ae70aeee68b84e41/airbyte_cdk/sources/declarative/transformations/transformation.py#L25

This means that the transformation is do in place and that the caller of this function should not store the result in a variable like https://github.com/airbytehq/airbyte/pull/48472/files#diff-6c388abccab93dc7577dc93da2d16f691f60bb9c806ec20011457fc10070b07cR708

It also probably means that we don't need to do a deep copy

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/components.py
Comment (maxi297) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/components.py:28]: If I understand correctly from the test, it does not only transform a record but also schemas, right? Can we update this description and re-assign the variable `record` to a variable with a better name to make it clear like `record_or_schema`?

**Review by maxi297 — COMMENTED** [review]
I've added a couple of comments but the core of it makes sense to me

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/components.py
Comment (brianjlai) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/components.py:13]: only nit is that can we include an explanation why the existing `AddFields` component was not compatible with hubspot to justify the need for a custom implementation. Hopefully our future selves thank us

**Review by brianjlai — COMMENTED** [review]
pending maxime's comment about the signature and in place transformation. lgtm!

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/components.py
Comment (pnilan) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/components.py:13]: Added. Good callout.

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/components.py
Comment (pnilan) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/components.py]: Yes, I agree with that.

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/components.py
Comment (pnilan) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/components.py]: Good catch -- updated accordingly.

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/components.py
Comment (brianjlai) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/components.py:32]: nit: the comment should mention its both records or the catalog schema

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/components.py
Comment (brianjlai) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/components.py:44]: Presumably this is a surrogate for detecting if this is a schema or a record. I think it would make it more clear if we assigned this to a variable for example: `is_schema = record_or_schema.get("properties") is not None`. And then in the following line: `if is_schema:`. And that way it works as a sort of self documenting piece of code

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/components.py
Comment (brianjlai) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/components.py:46]: I think the main thing that feels weird about this component/logic is that we're combining schema modification and record transformation into the same component. Its a little confusing in that regard, but not a dealbreaker.

But the reason why I'm a little opposed about this is that when we do move Hubspot to low-code, this custom component will only ever get invoked at runtime on records since thats how the low-code framework uses transformations. And we're going to have to reimplement this (or maybe we get it for free depending how the dynamic schema component is designed). So given that schema transformation should eventually be removed from the component, it makes me think that we should just implement the schema part separately from the custom component.

Again, I know we have a time crunch to get this done, so ultimately this is just a nit, but this was my main thought that kept popping up in my head during the review.

**Review by brianjlai — APPROVED** [review]
i think I still have some reservations about combining the `transform()` method to support both record and schema manipulation. vs separating them Mainly because it goes against what the intended use case of the `AddFields` component is supposed to be used. But also given the urgency, I don't need to strictly block on it.

approving, but pending regression test validation. I'd expect for it to fail since the schema is gonna have more fields than it used to, but we should check the results for unexpected failures

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/components.py
Comment (pnilan) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/components.py:32]: Agreed. Updated.

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/components.py
Comment (pnilan) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/components.py]: Actually the other way around, but yes updated to be self documenting -- if it has "properties" property it's the record, as these specific properties are nested under "properties".

**Pull request merged**
